### PR TITLE
ddl: Fix unstable `DROP TABLE`/`FLASHBACK TABLE`/`RECOVER TABLE`

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/8395, close https://github.com/pingcap/tiflash/issues/1664, close https://github.com/pingcap/tiflash/issues/3777

Problem Summary:

Problem 1:
There could be a chance that raft snapshot or raft command of a table comes after it has been dropped.
Because tiflash will ignore the previous `DROP TABLE`, because storage instance is not created yet. The storage instance is created as "non tombstone" after that when raft snapshot or raft command comes. And the storage instance will never be physically dropped until tiflash restarts. 

Problem 2:
In the second time when `syncTableSchema` calls `trySyncTableSchema` after table-id-mapping is up-to-date, it should use `mvcc get` to ensure it can use the table schema of tombstone table to create Storage instance or decode new raft logs, or some new columns data will not be decoded.
Previously `updateTiFlashReplica` will update the table info with the latest columns by accident. So tiflash pass the related tests sometimes.

```sql
CREATE TABLE t (a int, b int)
INSERT INTO t VALUES (1, 1);
ALTER TABLE t SET TIFLASH REPLICA 1;
# wait tiflash replica ready

ALTER TABLE t ADD COLUMN c int;
INSERT INTO t VALUES (1,2,3);
DROP TABLE t;

# the rows (1,2,3) comes after `drop table t` is executed in tiflash
RECOVER TABLE t; -- or FLASHBACK TABLE t

SELECT * FROM t;
# should return rows (1, 1, NULL), (1,2,3);
```

### What is changed and how it works?

This is a PR following https://github.com/pingcap/tiflash/pull/8421
The logic changes is these commits: https://github.com/pingcap/tiflash/pull/8422/files/d88f6b6f4ae4c8026b969cd9c5ae50924b179529..1fe991997055f755aea89cd2e2bdb8ab26a848bf

* In the first time when `syncTableSchema` calls `trySyncTableSchema`, it will not use `mvcc get` so that we can detect whether we need to update the table-id-mapping
* In the second time when `syncTableSchema` calls `trySyncTableSchema` after table-id-mapping is up-to-date, it will use `mvcc get` to ensure it can use the table schema of tombstone table to create Storage instance or decode new raft logs.
* `SchemaGetter::getTableInfoImpl` does not check the existence of `db_key`, so that we can still get the table info after database is dropped. (get ready for `FLASHBACK DATABASE ... TO ...`)
* If a storage instance is created with the table info got by `client-c mvcc get`, then we will create the storage instance with a tombstone timestamp. So that it can be physically dropped after GC time.
* `applySetTiFlashReplica` should only update the tiflash replica info instead of replacing all the table info, or it will lead to some later DDLs (changing partitions, etc) is not executed
* Simplify `recover table` codes using early return

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix a potential issue that some data can not be recovered by `RECOVER TABLE` and `FLASHBACK TABLE`
```

### CODE REVIEW

**Comment by CalvinNeo** [comment]
Can this pr fix https://github.com/pingcap/tiflash/issues/1664??

**Review by CalvinNeo — COMMENTED** [review]
Can this pr fix https://github.com/pingcap/tiflash/issues/1664 ? 
Seems the problem1 is https://github.com/pingcap/tiflash/issues/1664#issuecomment-808631344, and the problem2 is https://github.com/pingcap/tiflash/issues/1664#issue-842031800?

If so, please also close 1664.

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 1]
```diff
@@ -86,16 +91,28 @@ void SchemaBuilder<Getter, NameMapper>::applyCreateTable(DatabaseID database_id,
     }
 
     // If table is partition table, we will create the logical table here.
-    // Because we get the table_info, so we can ensure new_db_info will not be nullptr.
     auto new_db_info = getter.getDatabase(database_id);
-    applyCreateStorageInstance(new_db_info, table_info);
+    if (new_db_info == nullptr)
```
Comment (Lloyd-Pottiger) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:95]: Why is it called `new_db_info`?

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 2]
```diff
@@ -86,16 +91,28 @@ void SchemaBuilder<Getter, NameMapper>::applyCreateTable(DatabaseID database_id,
...
+        // the database has been dropped
+        LOG_INFO(
+            log,
+            "database is not exist in TiKV, may have been dropped, applyCreateTable is ignored, database_id={} "
+            "table_id={}",
+            database_id,
+            table_id);
+        return;
+    }
+    applyCreateStorageInstance(new_db_info, table_info, get_by_mvcc);
```
Comment (Lloyd-Pottiger) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:106]: add some comments to explain why creating logical table need to create storage instance.

**Comment by JaySon-Huang** [comment]
> Can this pr fix #1664 ? Seems the problem1 is [#1664 (comment)](https://github.com/pingcap/tiflash/issues/1664#issuecomment-808631344), and the problem2 is [#1664 (comment)](https://github.com/pingcap/tiflash/issues/1664#issue-842031800)?
> 
> If so, please also close 1664.

Yes, it can. Thanks for reminding

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp [hunk 3]
```diff
@@ -373,68 +391,100 @@ void SchemaBuilder<Getter, NameMapper>::applySetTiFlashReplica(DatabaseID databa
...
-            }
-            else
-            {
-                applyRecoverTable(db_info->id, table_id);
-            }
+            applyCreateTable(db_info->id, table_id);
         }
-        else
+        return;
+    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/TiDB/Schema/SchemaBuilder.cpp:407]:
```
1. create table t (I int);
2. alter table t partition by hash(id) partitions 8; <-- normal table to partition table
3. alter tiflash replica t set tiflash replica 1;
4. insert into t values (10001), (10002), ....; <-- insert data
```

What if the ordering come to tiflash is `3->1->2->4`?

Comment (JaySon-Huang) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp:407]: Does `3->1` mean the "Region snapshot" comes before the DDL `create table t` is executed in TiFlash?

Comment (Lloyd-Pottiger) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp:407]: yeah

1. tikv create normal table
2. tikv change the table to partition table
3. tikv send Region snapshot to tiflash
4. tiflash get `create table` DDL.

Comment (JaySon-Huang) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp:407]: When TiFlash try do decode the Region snapshot/raft command, it calls `AtomicGetStorageSchema`. Inside that function, if the storage instance is not yet created, it calls `syncTableSchema`. If `syncTableSchema` does not known the table id, it sync the DDL with tikv and apply `create table` in TiFlash.

https://github.com/pingcap/tiflash/blob/cf0e9affe7191b428a25142a91f240bce42823a1/dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp#L465-L488

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp (hunk 1 above)
Comment (JaySon-Huang) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp]: not important, name is changed

File: dbms/src/TiDB/Schema/SchemaBuilder.cpp (hunk 2 above)
Comment (JaySon-Huang) [reply dbms/src/TiDB/Schema/SchemaBuilder.cpp]: added

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM

File: tests/fullstack-test2/ddl/alter_drop_table.test [hunk 4]
```diff
@@ -44,66 +59,29 @@ mysql> set session tidb_isolation_read_engines='tiflash';select * from test.t3;
...
-
 mysql> drop table test.t3;
-
 => DBGInvoke __refresh_schemas()
-
 >> select tidb_database,tidb_name from system.tables where tidb_database = 'test' and tidb_name = 't3' and is_tombstone = 0
 
-
-## drop table arrive tiflash before ddl and insert, and do recover, check the data is not lost
-## because we want to test we actually drop the table, so please not use the same name for this table
```
Comment (JaySon-Huang) [inline tests/fullstack-test2/ddl/alter_drop_table.test:56]: This case is moved to `fullstack-test2/ddl/flashback`

**Review by hongyunyan — APPROVED** [review]
LGTM

**Pull request merged**